pkg/radius/handler/handler.go

```diff
@@ -121,44 +121,122 @@ func HandleRadiusAccessRequest(udpConn *net.UDPConn, tngfAddr, ueAddr *net.UDPAd
+			radiusLog.Error("[EAP] Received an EAP payload with code other than response. Drop the payload.")
+			return
+		}
+
+		eapTypeData := eap.EAPTypeData[0]
+		var eapExpanded *radius_message.EAPExpanded
+
+		switch eapTypeData.Type() {
+		case radius_message.EAPTypeExpanded:
+			eapExpanded = eapTypeData.(*radius_message.EAPExpanded)
+		default:
+			radiusLog.Errorf("[EAP] Received EAP packet with type other than EAP expanded type: %d", eapTypeData.Type())
+			return
+		}
+
+		if eapExpanded.VendorID != radius_message.VendorID3GPP {
+			radiusLog.Error("The peer sent EAP expended packet with wrong vendor ID. Drop the packet.")
+			return
+		}
+		if eapExpanded.VendorType != radius_message.VendorTypeEAP5G {
+			radiusLog.Error("The peer sent EAP expanded packet with wrong vendor type. Drop the packet.")
+			return
+		}
+
+		eap5GMessageID, anParameters, nasPDU, unmarshal_err := UnmarshalEAP5GData(eapExpanded.VendorData)
+		if unmarshal_err != nil {
+			radiusLog.Errorf("Unmarshalling EAP-5G packet failed: %+v", unmarshal_err)
+			return
+		}
+
+		if eap5GMessageID == radius_message.EAP5GType5GStop {
+			// Send EAP failure
+			// Build Radius message
+			responseRadiusMessage.BuildRadiusHeader(radius_message.AccessChallenge, message.PktID, message.Auth)
+			responseRadiusMessage.Payloads.Reset()
+
+			// EAP
+			identifier, random_err := GenerateRandomUint8()
+			if random_err != nil {
+				radiusLog.Errorf("Generate random uint8 failed: %+v", random_err)
 				return
 			}
+			responseRadiusPayload.BuildEAPfailure(identifier)
 
-			eapTypeData := eap.EAPTypeData[0]
-			var eapExpanded *radius_message.EAPExpanded
+			if requestMessageAuthenticator != nil {
+				tmpRadiusMessage := *responseRadiusMessage
+				payload := new(radius_message.RadiusPayload)
+				payload.Type = radius_message.TypeMessageAuthenticator
+				payload.Length = uint8(18)
+				payload.Val = make([]byte, 16)
 
-			switch eapTypeData.Type() {
-			case radius_message.EAPTypeExpanded:
-				eapExpanded = eapTypeData.(*radius_message.EAPExpanded)
-			default:
-				radiusLog.Errorf("[EAP] Received EAP packet with type other than EAP expanded type: %d", eapTypeData.Type())
-				return
+				tmpResponseRadiusPayload := responseRadiusPayload
+				tmpResponseRadiusPayload = append(tmpResponseRadiusPayload, *payload)
+
+				tmpRadiusMessage.Payloads = tmpResponseRadiusPayload
+
+				payload.Val = GetMessageAuthenticator(&tmpRadiusMessage)
+				responseRadiusPayload = append(responseRadiusPayload, *payload)
 			}
+			responseRadiusMessage.Payloads = responseRadiusPayload
 
-			if eapExpanded.VendorID != radius_message.VendorID3GPP {
-				radiusLog.Error("The peer sent EAP expended packet with wrong vendor ID. Drop the packet.")
-				return
+			// Send Radius message to UE
+			SendRadiusMessageToUE(udpConn, tngfAddr, ueAddr, responseRadiusMessage)
+			return
+		}
+
+		// Send Initial UE Message or Uplink NAS Transport
+		if session.ThisUE == nil { // if anParameters != nil {
```

Comment:
Does it need to check anParameters != nil

Reply:
UnmarshalEAP5GData returns a nil `anParameters` pointer only in case of an error during the unmarshalling process. We already handle unmarshalling error at line 155. It seems to be unnecessary but we can add it.

Reply:
Ok, so I think it's not necessary to check nil!
```suggestion
		if session.ThisUE == nil {
```